docs/reference/scripting/using.asciidoc

```diff
@@ -1,425 +1,471 @@
-    `source` as in the example above. A `stored` script is specified `id`
-    and is retrieved from the cluster state (see <<modules-scripting-stored-scripts,Stored Scripts>>).
-
+    Specifies the source of the script. An `inline` script is specified as `source` in the previous example. A `stored` script is specified by `id` and is retrieved from the cluster state using the `_scripts` endpoint.
 
 `params`::
 
     Specifies any named parameters that are passed into the script as
-    variables.
-
-[IMPORTANT]
-[[prefer-params]]
-.Prefer parameters
-========================================
+    variables. <<prefer-params,Use parameters>> instead of hard-coded values to decrease compile time.
 
-The first time Elasticsearch sees a new script, it compiles it and stores the
-compiled version in a cache. Compilation can be a heavy process.
-
-If you need to pass variables into the script, you should pass them in as
-named `params` instead of hard-coding values into the script itself.  For
-example, if you want to be able to multiply a field value by different
-multipliers, don't hard-code the multiplier into the script:
+[discrete]
+[[hello-world-script]]
+=== Write your first script
+<<modules-scripting-painless,Painless>> is the default scripting language
+for {es}. It is secure, performant, and provides a natural syntax for anyone
+with a little coding experience.
 
-[source,js]
-----------------------
-  "source": "doc['my_field'] * 2"
-----------------------
-// NOTCONSOLE
+A Painless script is structured as one or more statements and optionally
+has one or more user-defined functions at the beginning. A script must always
+have at least one statement.
 
-Instead, pass it in as a named parameter:
+The {painless}/painless-execute-api.html[Painless execute API] provides the ability to
+test a script with simple user-defined parameters and receive a result. Let's
+start with a simple script and review its constituent parts.
 
-[source,js]
-----------------------
-  "source": "doc['my_field'] * multiplier",
-  "params": {
-    "multiplier": 2
+[source,console]
+----
+POST /_scripts/painless/_execute <1>
+{
+  "script": { <2>
+    "source": "return 'Hello, world!';" <3>
   }
-----------------------
-// NOTCONSOLE
-
-The first version has to be recompiled every time the multiplier changes.  The
-second version is only compiled once.
+}
+----
+<1> Request URL
+<2> `script` object
+<3> `script` source
 
-If you compile too many unique scripts within a small amount of time,
-Elasticsearch will reject the new dynamic scripts with a
-`circuit_breaking_exception` error. For most contexts, you can compile up to 75
-scripts per 5 minutes by default. For ingest contexts, the default script
-compilation rate is unlimited. You can change these settings dynamically by
-setting `script.context.$CONTEXT.max_compilations_rate` eg.
-`script.context.field.max_compilations_rate=100/10m`.
+The request URL partially determines what script _contexts_ are available.
```

Comment:
What is this trying to convey?

Reply:
Mainly that the URL can vary, and that you can use scripts in a myriad of places outside of this particular script context. For example, you can include a script in a search context to calculate a value as part of a search, which is different from just running a script with the `POST` method like `POST /_scripts/painless/_execute`.

Reply:
Ahh ok.  Rather than the URL, it's really different features.  In a `_search` request, you may have use painless in the scoring context or the field context.

Reply:
Maybe this explanation is more clear:
>The request URL partially determines what script features are available. For example, you can use Painless in the
scoring context or the field context in a `_search` request. Contexts control how scripts run, what variables are available when the script runs, and what the return type is.

Reply:
Focusing on the URL confuses matters.

There are four different contexts in the scripted metric aggregation alone, which is also `_search`.

Reply:
Perhaps it's worth removing those sentences altogether?

Reply:
Yes, I see they're removed.  Thanks.